src/EllCrv/Misc.jl

```diff
@@ -129,13 +129,13 @@ function nrootscubic(b, c, d, p)
   fac = factor(f)
 
   if length(fac) == 1
-    if fac[first(keys(fac.fac))] == 3
+    if fac[first(fac)[1]] == 3
       return ZZ(3)
     else
       return ZZ(0)
     end
   elseif length(fac) == 2
-    if fac[first(keys(fac.fac))]== 1 && fac[first(keys(fac.fac))] == 1
+    if fac[first(fac)[1]]== 1 && fac[first(fac)[1]] == 1
```

Comment:
this line contains twice the same condition, doesn't it? Is this code maybe assuming that `first` modifies the state of the iterator?

Reply:
I would say this code is buggy: this function is meant to count roots of cubic polynomial modulo $p$. In this case there are exactly two factors... If there are two entries in `fac` then either both have multiplicity 1 (then one has degree 1 and the other degree 2). Or else one will have multiplicity 1 the other 2.

The comment in the next line says "one linear and one irreducible quadratic factor".
So I am sure what was really the intention here is to verify that the first and second/last entry both are. one. So it really should be
```suggestion
    if fac[first(fac)[1]]== 1 && fac[last(fac)[1]] == 1
```